### ISSUE TITLE
Path.GetDirectoryName slower on Linux than Windows Core

### ISSUE THREAD

**Issue by ianhays** [dotnet/corefx#3512]
The performance test being run:
```
[Benchmark]
[InlineData(10000)]
[InlineData(20000)]
[InlineData(30000)]
public void GetDirectoryName(int innerIterations)
{
    string testPath = PerfUtils.GetTestFilePath();
    foreach (var iteration in Benchmark.Iterations)
        using (iteration.StartMeasurement())
            for (int i = 0; i < innerIterations; i++)
            {
                Path.GetDirectoryName(testPath); Path.GetDirectoryName(testPath); Path.GetDirectoryName(testPath);
                Path.GetDirectoryName(testPath); Path.GetDirectoryName(testPath); Path.GetDirectoryName(testPath);
                Path.GetDirectoryName(testPath); Path.GetDirectoryName(testPath); Path.GetDirectoryName(testPath);
            }
}
```

Linux perf results (99.804 seconds):
```
<assemblies>
  <assembly name="System.Runtime.Extensions.Tests.dll" environment="64-bit .NET (unknown version) [collection-per-class, parallel (1 threads)]" test-framework="xUnit.net 2.1.0.3168" run-date="2015-09-29" run-time="06:05:42" total="3" passed="3" failed="0" skipped="0" time="99.804" errors="0">
    <errors />
    <collection total="3" passed="3" failed="0" skipped="0" name="Test collection for System.Runtime.Extensions.Tests.Perf_Path" time="96.597">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetDirectoryName(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetDirectoryName" time="16.1740961" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetDirectoryName(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetDirectoryName" time="32.1736083" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetDirectoryName(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetDirectoryName" time="48.2494629" result="Pass" />
    </collection>
  </assembly>
</assemblies>
```

Windows Core perf results (4.147 seconds):
```
<assemblies>
  <assembly name="System.Runtime.Extensions.Tests.dll" environment="64-bit .NET (unknown version) [collection-per-class, parallel (8 threads)]" test-framework="xUnit.net 2.1.0.3168" run-date="2015-09-28" run-time="15:19:54" total="3" passed="3" failed="0" skipped="0" time="4.327" errors="0">
    <errors />
    <collection total="3" passed="3" failed="0" skipped="0" name="Test collection for System.Runtime.Extensions.Tests.Perf_Path" time="4.147">
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetDirectoryName(innerIterations: 10000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetDirectoryName" time="1.2211779" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetDirectoryName(innerIterations: 20000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetDirectoryName" time="1.4310438" result="Pass" />
      <test name="System.Runtime.Extensions.Tests.Perf_Path.GetDirectoryName(innerIterations: 30000)" type="System.Runtime.Extensions.Tests.Perf_Path" method="GetDirectoryName" time="1.4950935" result="Pass" />
    </collection>
  </assembly>
</assemblies>
```

**Comment by stephentoub** [comment]
Is this still an issue after your HasIllegalCharacters change?

**Comment by ianhays** [comment]
Nope, this is fixed. I accidentally wrote "resolves #3515, #3515" instead of "Resolves #3515, #3512" on that PR.
